gridcomps/Cap/MAPL_CapGridComp.F90

```diff
@@ -1569,6 +1678,222 @@ subroutine rewind_clock(this, time, rc)
+        end if
+
+        call ESMF_GridCompRun(this%gcs(this%extdata_id), importState = this%child_imports(this%extdata_id), &
+           exportState = this%child_exports(this%extdata_id), &
+           clock = this%clock, userrc = status)
+        _VERIFY(status)
+
+        ! Call Record for intermediate checkpoint (if desired)
+        !  Note that we are not doing a Record for History.
+        ! ------------------------------------------------------
+        ! don't output reverse checkpoints
+        ! call ESMF_GridCompWriteRestart(this%gcs(this%root_id), importstate = this%child_imports(this%root_id), &
+        !      exportstate = this%child_exports(this%root_id), &
+        !      clock = this%clock_hist, userrc = status)
+        ! _VERIFY(status)
+
+        ! Advance the Clock before running History and Record
+        ! ---------------------------------------------------
+        if (.not. first) THEN
+           call ESMF_ClockAdvance(this%clock, rc = status)
+           _VERIFY(STATUS)
+           call ESMF_ClockAdvance(this%clock_hist, rc = status)
+           _VERIFY(STATUS)
+
+           ! Update Perpetual Clock
+           ! ----------------------
+
+           if (this%lperp) then
+              call Perpetual_Clock(this, status)
+              _VERIFY(status)
+           end if
+        end if
+
+        if (this%compute_throughput) then
+           call ESMF_VMBarrier(this%vm,rc=status)
+           _VERIFY(status)
+           this%starts%start_run_timer = MPI_WTime(status)
+        end if
+
+
+        call ESMF_ClockGet(this%clock, CurrTime = currTime, rc = status)
+        _VERIFY(status)
+        call ESMF_TimeGet(CurrTime, YY = AGCM_YY, &
+             MM = AGCM_MM, &
+             DD = AGCM_DD, &
+             H  = AGCM_H , &
+             M  = AGCM_M , &
+             S  = AGCM_S, rc=status)
+        _VERIFY(status)
+        if (this%AmIRoot) write(6,1000) AGCM_YY,AGCM_MM,AGCM_DD,AGCM_H,AGCM_M,AGCM_S
+1000    format(1x,'AGCM Date: ',i4.4,'/',i2.2,'/',i2.2,2x,'Time: ',i2.2,':',i2.2,':',i2.2)
+
+
+        _RETURN(_SUCCESS)
+
+     end subroutine
+
+     subroutine last_phase(rc)
+        integer, optional, intent(out) :: rc
+        integer :: status
+
+        if (this%compute_throughput) then
+           call ESMF_VMBarrier(this%vm,rc=status)
+           _VERIFY(status)
+           end_run_timer = MPI_WTime(status)
+        end if
+
+        call ESMF_GridCompRun(this%gcs(this%history_id), importstate=this%child_imports(this%history_id), &
+             exportstate = this%child_exports(this%history_id), &
+             clock = this%clock_hist, userrc = status)
+        _VERIFY(status)
+        ! Estimate throughput times
+        ! ---------------------------
+        if (this%compute_throughput) then
+           call print_throughput(rc=status)
+           _VERIFY(STATUS)
+        end if
+
+        _RETURN(_SUCCESS)
+
+     end subroutine
+
+     subroutine print_throughput(rc)
```

Comment:
The inner procedures here seem to serve a similar purpose to similarly named procedures elsewhere.     Would a bit of refactoring allow the same procedures to be used in both places?   I.e., can we eliminate some duplication with a modest effort?